process/block/preprocess/smartContractResults.go

```diff
@@ -0,0 +1,424 @@
+func (scr *smartContractResults) RemoveTxBlockFromPools(body block.Body, miniBlockPool storage.Cacher) error {
+	if body == nil {
+		return process.ErrNilTxBlockBody
+	}
+
+	err := scr.removeDataFromPools(body, miniBlockPool, scr.scrPool, block.SmartContractResultBlock)
+
+	return err
+}
+
+// RestoreTxBlockIntoPools restores the smartContractResults and miniblocks to associated pools
+func (scr *smartContractResults) RestoreTxBlockIntoPools(
+	body block.Body,
+	miniBlockHashes map[int][]byte,
+	miniBlockPool storage.Cacher,
+) (int, error) {
+	if miniBlockPool == nil {
+		return 0, process.ErrNilMiniBlockPool
+	}
+
+	scrRestored := 0
+	for i := 0; i < len(body); i++ {
+		miniBlock := body[i]
+		if miniBlock.Type != block.SmartContractResultBlock {
+			continue
+		}
+
+		strCache := process.ShardCacherIdentifier(miniBlock.SenderShardID, miniBlock.ReceiverShardID)
+		scrBuff, err := scr.storage.GetAll(dataRetriever.SmartContractResultUnit, miniBlock.TxHashes)
+		if err != nil {
+			return scrRestored, err
+		}
+
+		for txHash, txBuff := range scrBuff {
+			tx := smartContractResult.SmartContractResult{}
+			err = scr.marshalizer.Unmarshal(&tx, txBuff)
+			if err != nil {
+				return scrRestored, err
+			}
+
+			scr.scrPool.AddData([]byte(txHash), &tx, strCache)
+		}
+
+		err = scr.restoreMiniBlock(miniBlock, miniBlockPool, miniBlockHashes[i])
+		if err != nil {
+			return scrRestored, err
+		}
+
+		scrRestored += len(miniBlock.TxHashes)
+	}
+
+	return scrRestored, nil
+}
+
+// ProcessBlockSmartContractResults processes all the smartContractResult from the block.Body, updates the state
+func (scr *smartContractResults) ProcessBlockTransactions(body block.Body, round uint32, haveTime func() time.Duration) error {
+	// basic validation already done in interceptors
+	for i := 0; i < len(body); i++ {
+		miniBlock := body[i]
+		if miniBlock.Type != block.SmartContractResultBlock {
+			continue
+		}
+		if miniBlock.ReceiverShardID != scr.shardCoordinator.SelfId() {
+			continue
+		}
+
+		for j := 0; j < len(miniBlock.TxHashes); j++ {
+			if haveTime() < 0 {
+				return process.ErrTimeIsOut
+			}
+
+			txHash := miniBlock.TxHashes[j]
+			scr.scrForBlock.mutTxsForBlock.RLock()
+			txInfo := scr.scrForBlock.txHashAndInfo[string(txHash)]
+			scr.scrForBlock.mutTxsForBlock.RUnlock()
+			if txInfo == nil || txInfo.tx == nil {
+				return process.ErrMissingTransaction
+			}
+
+			currScr, ok := txInfo.tx.(*smartContractResult.SmartContractResult)
+			if !ok {
+				return process.ErrWrongTypeAssertion
+			}
+
+			err := scr.processSmartContractResult(
+				txHash,
+				currScr,
+				round,
+				miniBlock.SenderShardID,
+				miniBlock.ReceiverShardID,
+			)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// SaveTxBlockToStorage saves smartContractResults from body into storage
+func (scr *smartContractResults) SaveTxBlockToStorage(body block.Body) error {
```

Comment:
This method is not tested at all

Reply:
were copied from blockprocessor, not tested there with unit tests. Unit tests will be added. Added TODO

Reply:
ok